### PR TITLE
Add link to create new segment on send step [MAILPOET-5512]

### PR DESCRIPTION
**Pull request by johnolek**
## Description

This adds a link to create a new filter segment from the sending step. After creating a new segment, the user is returned to the send step with the filter segment pre-filled.

Originally included in https://github.com/mailpoet/mailpoet/pull/5141 but I closed that PR to create 2 instead.

I included a quick fix in this PR (f4752d3e20e4b2914407a98a97d954682c20c2d9) to ensure the filter segment tags show up in the listing for standard newsletters.

## Code review notes

If you can think of a simpler way of handling saving the email before the redirect I'm all ears.

I recommend hiding whitespace changes when reviewing because I wrapped some existing code in a React context and the diff with whitespace doesn't make sense.

## QA notes

_N/A_

## Linked PRs

_N/A_

## Linked tickets

https://mailpoet.atlassian.net/browse/MAILPOET-5512

## After-merge notes

_N/A_

## Tasks

- [X] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [X] I added sufficient test coverage
- [X] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes

### CODE REVIEW

File: mailpoet/assets/js/src/newsletters/send.tsx
Comment (johnolek) [reply mailpoet/assets/js/src/newsletters/send.tsx:871]: This isn't a change introduced by this PR (I think it's only being flagged because this part of the code changed indentation level) and I'm not entirely sure what the concern is. If it's about `wpPostId`, that value should be coming straight from the database and it's an integer column, so I don't think it's a risk.

**Comment by johnolek** [comment]
@rodrigoprimo I know I've just sent a lot of review requests your way. Please let me know if you'd like me to reassign any of them, though I'm hoping they shouldn't take too much time, especially with your level of familiarity with this part of the codebase.

File: mailpoet/assets/js/src/newsletters/send/filter-segment.tsx
Comment (rodrigoprimo) [inline mailpoet/assets/js/src/newsletters/send/filter-segment.tsx]: Is this attribute needed? If so, maybe we can use a variable for the URL, as it is the same as in line 167? It is just a detail, but I'm just trying to avoid a situation in the future where we update the URL in one place but don't update in the other creating a minor inconsistency.

Comment (rodrigoprimo) [reply mailpoet/assets/js/src/newsletters/send/filter-segment.tsx]: I wonder if there is a way to [`NEW_DYNAMIC_SEGMENT`](https://github.com/mailpoet/mailpoet/blob/2b5cf32b12226953b8ca2e04b1ffcc4e4eb76ea5/mailpoet/assets/js/src/segments/routes.ts#L1) here. Probably not, since it is a different React app.

File: mailpoet/assets/js/src/segments/dynamic/editor.tsx
Comment (rodrigoprimo) [inline mailpoet/assets/js/src/segments/dynamic/editor.tsx:55]: Instead of using the `onClick` event to return to the newsletter page, would it make sense to combine this logic with the logic above to define the value of `returnPage`? My first reaction without much thought is that returning to the newsletter page should be another possible value of `returnPage`.

**Review by rodrigoprimo — COMMENTED** [review]
Looks good to me! I left a few comments with small things for us to discuss.

> If you can think of a simpler way of handling saving the email before the redirect I'm all ears.

The way you implemented it makes sense to me. That being said, I don't think I'm the best person to tell whether or not there is a better approach. So my plan is to ask for a second review specifically about this after we discuss the points that I raised.

File: mailpoet/assets/js/src/segments/dynamic/editor.tsx
Comment (johnolek) [reply mailpoet/assets/js/src/segments/dynamic/editor.tsx:55]: That's what I wanted to do as well, but I don't think it's possible because of how this component is wrapped in a `HashRouter`. Since the root of this react app is essentially `admin.php?page=mailpoet-segments#`, I don't think there's a way to specify a `returnPage` that would navigate to a different page.

File: mailpoet/assets/js/src/newsletters/send/filter-segment.tsx
Comment (johnolek) [reply mailpoet/assets/js/src/newsletters/send/filter-segment.tsx]: Details are important! Thank you for this comment. I extracted it to a variable in f6ac44a0c1690ec10d3d00d1daad330593baed31.

It wouldn't be strictly necessary to have the `href` since the click event should always get intercepted, but without the `href` the browser doesn't see it as an interactive element, so it doesn't exist in the tab order among other issues.

File: mailpoet/assets/js/src/segments/dynamic/editor.tsx
Comment (rodrigoprimo) [reply mailpoet/assets/js/src/segments/dynamic/editor.tsx:55]: Fair point, thanks for looking into it.

**Review by rodrigoprimo — DISMISSED** [review]

**Comment by rodrigoprimo** [comment]
I'm assigning this PR to @JanJakes for a second look, specially around the part below as I'm not sure if there is a better approach.

> If you can think of a simpler way of handling saving the email before the redirect I'm all ears.

File: mailpoet/assets/js/src/newsletters/send.tsx
Comment (JanJakes) [inline mailpoet/assets/js/src/newsletters/send.tsx]: The issue is that you're creating a new object for each render. Any solution that can prevent that will be fine. I don't know much about class components, but probably sth like `this.value`, where `value = { saveDraftNewsletter: this.saveDraftNewsletter };` could do it? At least in case it's only `render` that's called again and again.

**Comment by JanJakes** [comment]
@johnolek @rodrigoprimo Could you please point me to what is the problem, exactly? I found a few places with a redirect but without additional context, I'm not sure what is the issue. Is it the fact that React context is used? Or the "what to do in class component" when `useMemo` is recommended (added a comment)?

File: mailpoet/assets/js/src/newsletters/send.tsx
Comment (JanJakes) [inline mailpoet/assets/js/src/newsletters/send.tsx]: Could this be an async function instead? Passing callback seems a bit old-school, but maybe we're using old code here, that is not easy to wrap with a promise...?

**Comment by johnolek** [comment]
> @johnolek @rodrigoprimo Could you please point me to what is the problem, exactly? I found a few places with a redirect but without additional context, I'm not sure what is the issue. Is it the fact that React context is used? Or the "what to do in class component" when `useMemo` is recommended (added a comment)?

@JanJakes Thanks for taking a look at this! There's not necessarily a problem. My main question was about the React context, specifically if there's a more straightforward or more idiomatic way to achieve the same behavior, basically if any of this would be violating best practices. I was wondering because it doesn't seem like we use contexts in many other places in the codebase and I haven't used them before myself.

File: mailpoet/assets/js/src/newsletters/send.tsx
Comment (johnolek) [reply mailpoet/assets/js/src/newsletters/send.tsx]: In 58978ec921db86709636ff4cc03ebc4c6c7cdd49 I changed this so the value exists in the component's state. Does that seem reasonable to you?

File: mailpoet/assets/js/src/newsletters/send.tsx
Comment (johnolek) [reply mailpoet/assets/js/src/newsletters/send.tsx]: Thanks for pointing this out! I updated this in ee7e9fee7a7bd971c888cb57d70aafdff438c7a8. Let me know if it looks off in any way.

**Comment by JanJakes** [comment]
@johnolek Ahh, the question "whether to use context" here is probably related to options of solving it without it. I quickly checked, and given that it's needed in a field that's initialized using `component: withBoundary(FilterSegment)`, props are not a good way here. As for a function that could be just imported, that doesn't work easily either, considering that `this.something` calls are needed. Then context seems fine to me 👍 LGTM, feel free to move it forward.

File: mailpoet/assets/js/src/newsletters/send.tsx
Comment (JanJakes) [reply mailpoet/assets/js/src/newsletters/send.tsx]: @johnolek Yeah, that looks good. Another good idea how to make the object identity stable 👍

File: mailpoet/assets/js/src/newsletters/send/filter-segment.tsx
Comment (JanJakes) [inline mailpoet/assets/js/src/newsletters/send/filter-segment.tsx]: The only tiny nitpick is I'd personally convert `await something().then(...)` to:

```ts
await something();
// here do the "then"
```

**Review by JanJakes — DISMISSED** [review]

File: mailpoet/assets/js/src/newsletters/send/filter-segment.tsx
Comment (johnolek) [reply mailpoet/assets/js/src/newsletters/send/filter-segment.tsx]: I always appreciate "nitpick"s! I modified the commit so now 2a1eff63eb56a4419d8f63ec0352410086043f82 has it like you describe.

**Comment by johnolek** [comment]
Thanks @JanJakes for your input/confirmation! I made one small change to address your [comment](https://github.com/mailpoet/mailpoet/pull/5176#discussion_r1356934124). Could you (or @rodrigoprimo) please approve again when you have a chance?

**Review by JanJakes — APPROVED** [review]

**Comment by Aschepikov** [comment]
HI @johnolek can you rebase it pls to the current master

**Comment by johnolek** [comment]
@Aschepikov Rebased and pushed, let me know if you need anything else!

**Comment by johnolek** [comment]
Sorry @Aschepikov, I just pushed again because I didn't have all the latest changes from `trunk` locally the last time I rebased. This branch should now be fully up to date.

**Review by Aschepikov — APPROVED** [review]
GJ!

**Pull request merged**
